### ISSUE TITLE
Parse markdown code-cell outputs as part of the document AST

### ISSUE THREAD

**Issue by choldgraf** [jupyter-book/mystmd#1026]
Jupyter notebook cells can produce text output that is specific to markdown (`text/markdown`). We should parse this output along with the rest of the markdown in the notebook. This would allow for **programmatic generation of notebook content**, and along with the `{embed}` directive could allow you to nicely stitch together MyST content via any Jupyter Kernel.

### An example from Jupyter Book 1

For example:

<img src="https://github.com/executablebooks/mystmd/assets/1839645/1bdd6f58-d437-4c63-bdd5-03c173e2b3d9" width=300 />

As an example, [the MyST-nb Sphinx extension documents this functionality here](https://myst-nb.readthedocs.io/en/latest/render/format_code_cells.html#markdown).

### An example from our docs where this would be useful

In the admonitions docs here:

https://github.com/executablebooks/mystmd/blob/c0f51ee51890158081736722cb01bcf510699786/docs/admonitions.md?plain=1#L34-L98

We have a big list of all the different types of admonitions for demonstration. It's just boilerplate myst repeated over and over. If we parsed Markdown outputs as MyST, we could replace it with something like:

``````python
inner = []
for admonition in ["note", "tip", "warning"]:
	s = """
	````{tab-item}
    ```{%s}
    This is an %s admonition. 
    ```
    ````
    """ % admonition
	inner.append(s)
template = """
`````{tab-set}
%s
````` % "\n".join(inner)
"""
``````

### Related issues

- https://github.com/jupyter-book/mystmd/issues/1633 is about generating AST directly rather than `text/markdown`

**Comment by HelgeGehring** [comment]
continuing #1559 here:

An example for this would be the following:

- Right now I have a nice notebook which analyzes exactly one physical structure.
- The analysis results in several graphs / tables / images / ..., each generated by a code cell

My problem now is, that I have 2-20 different structures. For each I get graphs based on the same code.
Right now the only solution is to put them all below each other, leading to a quite long doc.

Better would be to have
- tabs in which I can select which structure I want to see

and/or

- a table which has only the main results (maybe broken down in a number or two), but on click on a row, I see a detailed analysis for that structure.


A simplified example might be statistics about bank accounts (not scientific, but less abstract).
- It's simple to write a notebook which produces graphs / tables, ... based on an account number
- Now I have 5 accounts and want this one notebook to do the same analysis with all of them. An overview would be the balance of each one in a table. But I'd like to be able to switch tabs/expand rows of the different account to see the detailed graphs.


For different use-cases different approaches might be best. I'm wondering what might be possible:
- have on the complete notebook level tabs, i.e. execute the notebook like github actions matrix with different parameters
- have tabs within each output (which might be syncronized when switching (?))
- Having outputs combined with hidden content, e.g. a table in which i can expand individual rows to see more details.

**Comment by choldgraf** [comment]
## Workaround: Write to temporary text files and then use `{include}`

I discovered this workaround today, it's a bit cludgy but isn't too hacky. It takes advantage of the fact that **code cells are executed before a page is parsed by MyST**. This means that you can do something like the following:

````
## Generate content with Jupyter

```{code-cell} python
from pathlib import Path
p = Path("../_build/txt/tmp.txt")
p.parent.mkdir(parents=True, exist_ok=True)
_ = p.write_text("- **Testing**\n- Testing two\n- Testing three")
```

And then include it in the page with MyST markdown like so:

```{include} ../_build/txt/tmp.txt
```
````

This will:

1. Generate some MyST Markdown in Jupyter
2. Write it to a `.txt` file
3. And later in the page, in MyST MD, we reference that `.txt` file with an `{include}` statement

So the file is first executed, the `txt` file is created, and MyST then includes it

**Comment by agoose77** [comment]
@choldgraf we should also make it possible to `include` files with `.myst.json` extensions, to support pulling in AST.

**Comment by choldgraf** [comment]
Yeah I was thinking that too. It also made me wonder if the notebook cell MyST support could be more like the plugin structure, rather than making it language specific.

For example, a cell tag like `output-myst` or `output-myst-ast` that would tell MyST to parse stdout as MyST (similar to what executable plugins do).

Does that make sense? If so I can update the issue body to reflect that suggestion.

**Comment by agoose77** [comment]
@choldgraf the natural way to do this would be to define a MIME type for MyST AST, and recognise it in our transforms!

**Comment by choldgraf** [comment]
That sounds like a good idea. Though if it were the only way,  then each kernel would need to have a package that outputs MyST right? The benefit of tags and stdout is that anybody in any language could use it without developing anything specific.

**Comment by agoose77** [comment]
Although we don't need a package for this (certainly with ipython, you can just use `display`), I'm curious to understand your thought process - are you picturing a user printing stringified json to the stdout? Or myst markup, I.e text/markdown?

**Comment by choldgraf** [comment]
My idea was inspired by the way that you handled "black box" outputs for the executable plugins infrastructure. So below I'll share a Python and an R cell that would generate either MyST MD or MyST AST. At build time, if the tag was identified on a cell, then any `text/plain` output (or however stdout gets logged) would be parsed differently. Without those tags, the output would just be parsed like any other stdout.

Python - MyST MD

````
```{code-cell} python
:tags: output-myst-md
print("- **Bolded** list item")
```
````

Python - MyST AST

````
```{code-cell} python
:tags: output-myst-ast
ast = {
      "type": "list",
      "ordered": false,
      "spread": false,
      "children": [
        {
          "type": "listItem",
          "spread": True,
          "children": [
            {
              "type": "strong",
              "children": [
                {
                  "type": "text",
                  "value": "Bolded"
                }
              ]
            },
            {
              "type": "text",
              "value": " list item"
            }
          ]
        }
      ]
    }
print(ast)
```
````

**Comment by fperez** [comment]
I love the idea of parsing plain/MysT mardkown proper, as it's very easy to generate from many tools. In my code I very often have something like

```python
from IPython.display import display, Markdown
md = lambda s: display(Markdown(s))
```

and I use `md(x)` everywhere as a "print markdown" shorthand to generate legible, pretty reporting with zero fuss.

Having the ability to properly handle that (along perhaps with a new `MyST` object in `IPython.display`) would be very useful, I think.

**Comment by choldgraf** [comment]
I decided to create a separate issue to track generating MyST AST directly from cell outputs, since that might be an easier short-term solution and get us part of the way there:

- https://github.com/jupyter-book/mystmd/issues/1633

**Comment by agoose77** [comment]
@rowanc1 @stevejpurves @fwkoch @choldgraf

I can't seem to paste these in Discord right now. Here are some quick discussion videos (~2min each) about the work I've been doing on Markdown output parsing:

https://github.com/user-attachments/assets/e0d61b90-2148-42e5-b433-1fddbad6f301
https://github.com/user-attachments/assets/a2d69933-1d2b-4c2a-8c83-3aba3b8f7c03

The PR I'm most excited about is currently https://github.com/jupyter-book/mystmd/pull/1671.

Need to dash, no rush here, but it would be good to book in a video meeting at some point.

**Comment by fperez** [comment]
I'm very far from an expert in the nuances of the AST-consuming tools that may already exist, but FWIW I found your argument compelling, @agoose77 (and thanks for those explanatory videos!). 

A mechanism to tag "escape hatches" that transparently allow for third parties to opt-in with their own implementation and otherwise pass through silently is pragmatic and useful (I seem to recall the C++ standards group having a similar mechanism for compiler extensions, though my memory could be off).

It gives room for experimenting with tradeoffs and side effects before settling into a solution that gets baked into the AST proper, which seems like a very important benefit. Thanks!

**Comment by fperez** [comment]
Hi folks - came to the tracker looking for this topic, only to find not only there was already an issue, but I'd even participated :)

I'm now using the execution machinery a lot more, both in md files and in notebooks, and first - it rocks! But the lack of proper inclusion of Markdown output into the generated docs is a bigger bummer than I'd originally realized, so I'm here to advocate for prioritizing this much, much higher.

Here's my argument: I've been thinking a lot about Obsidian-like functionality in MyST workflows, and one of Obsidian's most magic-like powers is the ability to effortlessly drop a tiny bit of SQL-like code and get instantly massive amounts of rich, well formatted output about a vault, using the [Dataview plugin](https://obsidian.rocks/dataview-in-obsidian-a-beginners-guide/).

In the long run, I think it would be a _very_ fun project to imagine building something akin to a dataview kernel that would operate on all the data indexed in a MyST project. But that's not what I'm advocating for here, yet. 

I simply would love to see proper support for displaying MD output straight from both md and ipynb files executed at build time. That would unlock a huge amount of data-driven functionality that users could develop on their own, by having md or ipynb files (at their preference) that do things like compiling tables of publications, dynamic calendars (at build time), summaries about other files on the site, etc.

Being able to interactively write and test such files as live notebooks with Jupyter, and then have the results properly rendered on the site, would be a fluid, lovely workflow. Today it _almost_ works, except that something like this in a notebook:

<img width="488" alt="Image" src="https://github.com/user-attachments/assets/95a158a5-0551-4dee-8f97-a5154b273930" />

unfortunately results in output like this (I had tagged the cell with `remove-input` so the code would be invisible on the resulting page):

<img width="157" alt="Image" src="https://github.com/user-attachments/assets/4d80ea0b-8a84-46f8-a336-c179a258c03e" />

As an added benefit, having this handling of md outputs fixed is a necessary step for the more ambitious Obsidian-like idea, so we can fix something that will unlock a huge amount of potential, and then perhaps some enterprising soul will undertake that fun endeavor.

As always, thanks team for the great work!

**Comment by agoose77** [comment]
@fperez thanks for revisiting the issue! This feature is tracked on our for-beta project category, and we're moving towards it incrementally with #1900 and #1903. Keep checking back as we make headway!
